File: apps/web/app/utils/publicPage.ts
// The owner's public page title and emoji favicon. These mirror the server's
// internal/publicpage checks so the Publish dialog can explain an issue before
// saving; the server stays authoritative
// (docs/adr/0042-public-page-title-and-favicon.md).

export const MAX_TITLE_GRAPHEMES = 70;
const MAX_TITLE_CODE_POINTS = 560;
const MAX_FAVICON_CODE_POINTS = 16;

export type PublicPageIssue
  = 'too_long' | 'invalid_characters' | 'invalid_emoji';

export interface NormalizedPublicPageField {
  /** The value the server stores, or null when the field clears. */
  readonly value: string | null;
  /** Every issue, sorted; empty when the value is valid. */
  readonly codes: PublicPageIssue[];
}

const EDGE_WHITE_SPACE = /^\p{White_Space}+|\p{White_Space}+$/gu;
// Control and format characters, except U+200D ZERO WIDTH JOINER.
const INVISIBLE = /\p{Cc}|(?!\u200D)\p{Cf}/u;
const PICTOGRAPHIC = /^\p{Extended_Pictographic}$/u;
const REGIONAL_INDICATOR = /^\p{Regional_Indicator}$/u;
const UNRESERVED = /[A-Za-z0-9\-._~]/u;

const graphemes = (text: string): string[] => [
  ...new Intl.Segmenter('und', { granularity: 'grapheme' }).segment(text),
].map(({ segment }) => segment);

// What may follow an emoji base inside one sequence: ZWJ, variation
// selectors, skin-tone modifiers, and tag characters.
const emojiComponent = (codePoint: number): boolean =>
  codePoint === 0x200d || codePoint === 0xfe0e || codePoint === 0xfe0f
  || (codePoint >= 0x1f3fb && codePoint <= 0x1f3ff)
  || (codePoint >= 0xe0020 && codePoint <= 0xe007f);

const emojiCluster = (cluster: string): boolean => {
  const characters = [...cluster];
  if (
    characters.length === 2
    && characters.every((character) => REGIONAL_INDICATOR.test(character))
  ) {
    return true;
  }
  const [base, ...rest] = characters;
  if (base === undefined || !PICTOGRAPHIC.test(base)) return false;
  return rest.every((character) =>
    PICTOGRAPHIC.test(character) || emojiComponent(character.codePointAt(0)!));
};

export function normalizePublicTitle(raw: string): NormalizedPublicPageField {
  const title = raw.replace(EDGE_WHITE_SPACE, '');
  if (title === '') return { value: null, codes: [] };
  const codes: PublicPageIssue[] = [];
  if (INVISIBLE.test(title)) codes.push('invalid_characters');
  if (
    [...title].length > MAX_TITLE_CODE_POINTS
    || graphemes(title).length > MAX_TITLE_GRAPHEMES
  ) {
    codes.push('too_long');
  }
  return codes.length === 0
    ? { value: title, codes }
    : { value: null, codes: codes.sort() };
}

export function normalizeFaviconEmoji(raw: string): NormalizedPublicPageField {
  const emoji = raw.replace(EDGE_WHITE_SPACE, '');
  if (emoji === '') return { value: null, codes: [] };
  const clusters = graphemes(emoji);
  if (
    [...emoji].length > MAX_FAVICON_CODE_POINTS
    || clusters.length !== 1
    || !emojiCluster(emoji)
  ) {
    return { value: null, codes: ['invalid_emoji'] };
  }
  return { value: emoji, codes: [] };
}

/** The favicon data: URL the public page uses, byte for byte the server's. */
export function faviconHref(emoji: string): string {
  const svg = '<svg xmlns=\'http://www.w3.org/2000/svg\' viewBox=\'0 0 100 100\'>'
    + `<text y='.9em' font-size='90'>${emoji}</text></svg>`;
  let encoded = '';
  for (const byte of new TextEncoder().encode(svg)) {
    const character = String.fromCharCode(byte);
    encoded += byte < 0x80 && UNRESERVED.test(character)
      ? character
      : `%${byte.toString(16).toUpperCase().padStart(2, '0')}`;
  }
  return `data:image/svg+xml,${encoded}`;
}

File: apps/server/internal/publicpage/publicpage.go
// Package publicpage validates and renders the owner-set public page title
// and emoji favicon, publication settings stored beside the slug. See
// docs/adr/0042-public-page-title-and-favicon.md.
package publicpage

//go:generate go run extpict_gen.go

import (
	"sort"
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/rivo/uniseg"
)

const (
	// MaxTitleGraphemes bounds the title as a reader sees it.
	MaxTitleGraphemes = 70
	// MaxTitleRunes bounds its storage, since one grapheme may hold many
	// code points; the resumes_public_title_check constraint matches.
	MaxTitleRunes = 560
	// MaxFaviconRunes bounds one emoji grapheme. The longest standard
	// sequences, family ZWJ sequences with skin tones, use 11 code points.
	MaxFaviconRunes = 16
)

// Issue codes reported under the publish request's field paths.
const (
	// CodeTooLong reports a title over MaxTitleGraphemes or MaxTitleRunes.
	CodeTooLong = "too_long"
	// CodeInvalidCharacters reports a control or format character in a title.
	CodeInvalidCharacters = "invalid_characters"
	// CodeInvalidEmoji reports a favicon that is not exactly one emoji.
	CodeInvalidEmoji = "invalid_emoji"
)

// Result is one normalized field. Value is nil when the field clears; Codes
// lists every issue, sorted, and is empty when the value is valid.
type Result struct {
	Value *string
	Codes []string
}

// NormalizeTitle trims Unicode white space and validates a public title.
func NormalizeTitle(raw string) Result {
	title := trim(raw)
	if title == "" {
		return Result{}
	}
	var codes []string
	for _, r := range title {
		if unicode.Is(unicode.Cc, r) || (unicode.Is(unicode.Cf, r) && r != zeroWidthJoiner) {
			codes = append(codes, CodeInvalidCharacters)
			break
		}
	}
	if utf8.RuneCountInString(title) > MaxTitleRunes || uniseg.GraphemeClusterCount(title) > MaxTitleGraphemes {
		codes = append(codes, CodeTooLong)
	}
	if len(codes) != 0 {
		sort.Strings(codes)
		return Result{Codes: codes}
	}
	return Result{Value: &title}
}

// NormalizeFaviconEmoji trims Unicode white space and accepts exactly one
// emoji grapheme: an Extended_Pictographic base with only emoji modifiers,
// variation selectors, ZWJ, and tag characters after it, or one flag of two
// Regional Indicators.
func NormalizeFaviconEmoji(raw string) Result {
	emoji := trim(raw)
	if emoji == "" {
		return Result{}
	}
	if utf8.RuneCountInString(emoji) > MaxFaviconRunes || uniseg.GraphemeClusterCount(emoji) != 1 || !emojiCluster(emoji) {
		return Result{Codes: []string{CodeInvalidEmoji}}
	}
	return Result{Value: &emoji}
}

// EffectiveTitle is the page's <title>: the owner's title, or the default.
func EffectiveTitle(publicTitle *string, fullName string) string {
	if publicTitle != nil && *publicTitle != "" {
		return *publicTitle
	}
	return fullName + " — Resume"
}

// FaviconHref is the exact data: URL of an emoji favicon. Every byte outside
// the URL-unreserved set is percent-encoded, so the value cannot leave its
// attribute or the SVG text node whatever the emoji holds.
func FaviconHref(emoji string) string {
	svg := "<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 100 100'>" +
		"<text y='.9em' font-size='90'>" + emoji + "</text></svg>"
	var out strings.Builder
	out.WriteString("data:image/svg+xml,")
	for index := range len(svg) {
		b := svg[index]
		if unreserved(b) {
			out.WriteByte(b)
			continue
		}
		out.WriteByte('%')
		out.WriteByte(upperHex[b>>4])
		out.WriteByte(upperHex[b&0x0F])
	}
	return out.String()
}

const (
	zeroWidthJoiner = 0x200D
	upperHex        = "0123456789ABCDEF"
)

func trim(raw string) string {
	return strings.TrimFunc(raw, func(r rune) bool { return unicode.Is(unicode.White_Space, r) })
}

func unreserved(b byte) bool {
	return b >= 'A' && b <= 'Z' || b >= 'a' && b <= 'z' || b >= '0' && b <= '9' ||
		b == '-' || b == '.' || b == '_' || b == '~'
}

func emojiCluster(cluster string) bool {
	runes := []rune(cluster)
	if len(runes) == 2 && regionalIndicator(runes[0]) && regionalIndicator(runes[1]) {
		return true
	}
	if !isExtendedPictographic(runes[0]) {
		return false
	}
	for _, r := range runes[1:] {
		if !isExtendedPictographic(r) && !emojiComponent(r) {
			return false
		}
	}
	return true
}

func regionalIndicator(r rune) bool { return r >= 0x1F1E6 && r <= 0x1F1FF }

// emojiComponent covers what may follow an emoji base inside one sequence:
// ZWJ, text and emoji variation selectors, skin-tone modifiers, and tags.
func emojiComponent(r rune) bool {
	return r == zeroWidthJoiner || r == 0xFE0E || r == 0xFE0F ||
		(r >= 0x1F3FB && r <= 0x1F3FF) || (r >= 0xE0020 && r <= 0xE007F)
}

func isExtendedPictographic(r rune) bool {
	index := sort.Search(len(extendedPictographic), func(i int) bool { return extendedPictographic[i][1] >= r })
	return index < len(extendedPictographic) && extendedPictographic[index][0] <= r
}
